File: 백준/Gold/19238. 스타트 택시/스타트 택시.py
import sys
input = sys.stdin.readline
from collections import deque

n, m, fuel = map(int, input().split())
graph = [list(map(int,input().split())) for _ in range(n)]
tx, ty = list(map(int, input().split()))
guest = [list(map(int, input().split())) for _ in range(m)]

tx, ty = tx - 1, ty - 1
for i in range(m):
    x, y, nx, ny = guest[i]
    guest[i] = [x-1, y-1, nx-1, ny-1]
guest.sort()
dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

def choose_guest(x, y, fuel):
    queue = deque([(x,y)])
    visited = [[0]*n for _ in range(n)]
    min_distance = sys.maxsize
    candiate = []
    while queue:
        x, y = queue.popleft()
        if fuel < visited[x][y]:
            return -1, -1
        if visited[x][y] > min_distance:
            break
        if people[x][y] > -1:
            min_distance = visited[x][y]
            candiate.append([x,y])
            continue
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if nx < 0 or ny < 0 or nx >= n or ny >= n:
                continue
            if visited[nx][ny] or graph[nx][ny] == 1:
                continue
            visited[nx][ny] = visited[x][y] + 1
            queue.append((nx,ny))
    if candiate:
        candiate.sort()
        x, y = candiate[0]
        return people[x][y], visited[x][y]
    return -1, -1

def go_destination(x, y, fuel, gx, gy):
    queue = deque([(x,y)])
    visited = [[0]*n for _ in range(n)]
    while queue:
        x, y = queue.popleft()
        if fuel < visited[x][y]:
            return -1

        if gx == x and gy == y:
            return visited[x][y]

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if nx < 0 or ny < 0 or nx >= n or ny >= n:
                continue
            if visited[nx][ny] or graph[nx][ny] == 1:
                continue
            visited[nx][ny] = visited[x][y] + 1
            queue.append((nx,ny))
    return -1

while True:
    if not guest:
        print(fuel)
        break

    people = [[-1]*n for _ in range(n)]
    for i, cur_guest in enumerate(guest):
        x, y, _, _ = cur_guest
        people[x][y] = i

    num, spent_fuel = choose_guest(tx, ty, fuel)
    if num == -1:
        print(-1)
        break
    
    # 출발지로 가기
    fuel -= spent_fuel
    tx, ty = guest[num][0], guest[num][1]

    # 목적지로 가기
    spent_fuel = go_destination(tx, ty, fuel, guest[num][2], guest[num][3])
    if spent_fuel == -1:
        print(-1)
        break
    
    fuel += spent_fuel

    tx, ty = guest[num][2], guest[num][3]

    # 손님 삭제
    guest.pop(num)

File: 백준/Gold/19238. 스타트 택시/스타트 택시.js
const input = require("fs")
  .readFileSync(
    process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt"
  )
  .toString()
  .split("\n");

let index = 0;
const [n, m, init_fuel] = input[index++].split(" ").map(Number);
let fuel = init_fuel;

const board = [];
for (let i = 0; i < n; i++) {
  board.push(input[index++].split(" ").map(Number));
}

let taxi = input[index++].split(" ").map((num) => Number(num) - 1);

const passenger = [];
for (let i = 0; i < m; i++) {
  passenger.push(input[index++].split(" ").map((num) => Number(num) - 1));
}

const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];

const MAX_DISTANCE = 20 * 20 + 1;

console.log(move_taxi());

function move_taxi() {
  for (let t = 0; t < m; t++) {
    const [start_distance, passenger_index] = get_passenger();

    if (fuel < start_distance || start_distance === MAX_DISTANCE) {
      return -1;
    }

    const [start_x, start_y, end_x, end_y] = passenger[passenger_index];

    fuel -= start_distance;
    taxi = [start_x, start_y];

    const distance = get_distance(taxi[0], taxi[1]);
    const end_distance = distance[end_x][end_y];

    if (fuel < end_distance || end_distance === MAX_DISTANCE) {
      return -1;
    }

    fuel += end_distance;
    passenger[passenger_index] = [];
    taxi = [end_x, end_y];
  }

  return fuel;
}

function get_passenger() {
  const candidate = [];
  const distance = get_distance(taxi[0], taxi[1]);

  for (let i = 0; i < m; i++) {
    if (!passenger[i].length) continue;

    const [x, y, _, __] = passenger[i];
    candidate.push([distance[x][y], x, y, i]);
  }

  candidate.sort((a, b) => {
    if (a[0] === b[0]) {
      if (a[1] === b[1]) {
        return a[2] - b[2];
      }
      return a[1] - b[1];
    }
    return a[0] - b[0];
  });

  const [start_distance, _, __, passenger_index] = candidate[0];
  return [start_distance, passenger_index];
}

function get_distance(x, y) {
  const queue = [[x, y]];
  const visited = Array.from({ length: n }, () => Array(n).fill(MAX_DISTANCE));
  visited[x][y] = 0;

  while (queue.length) {
    const [x, y] = queue.shift();

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
      if (board[nx][ny] === 1) continue;
      if (visited[nx][ny] !== MAX_DISTANCE) continue;

      queue.push([nx, ny]);
      visited[nx][ny] = visited[x][y] + 1;
    }
  }

  return visited;
}
